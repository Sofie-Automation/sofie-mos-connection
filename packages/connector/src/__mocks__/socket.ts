/* eslint-disable @typescript-eslint/unbound-method */
import { EventEmitter } from 'events'
import { Socket } from 'net'

import * as iconv from 'iconv-lite'
import { getMessageId } from '../__tests__/lib.js'
import { getMosTypes } from '@mos-connection/model'
import { Mocked, vitest } from 'vitest'
iconv.encodingExists('utf16-be')

const mosTypes = getMosTypes(true)

// Mock the Socket class in 'net':
const setTimeoutOrg = setTimeout

const instances: Array<ISocketMock> = []

export class SocketMock extends EventEmitter implements Socket {
	// Mock implementation:
	bufferSize = 0
	remoteAddress = ''
	remoteFamily = ''
	remotePort = 0
	localAddress = ''
	localPort = 0
	bytesRead = 0
	bytesWritten = 0
	connecting = false
	destroyed = false
	writable = false
	readable = false

	readonly readableEncoding: BufferEncoding | null = null
	readonly readableEnded: boolean = false
	readonly readableFlowing: boolean | null = null

	readonly writableEnded: boolean = false
	readonly writableFinished: boolean = false
	readonly writableHighWaterMark: number = 0
	readonly writableLength: number = 0
	readonly writableObjectMode: boolean = false

	readonly readableHighWaterMark: number = 0
	readonly readableLength: number = 0
	readonly readableObjectMode: boolean = false

	public name = ''
	public connectedPort = 0
	public connectedHost = ''

	// [Symbol.asyncIterator](): AsyncIterableIterator<any>

	private _responses: Array<ReplyTypes> = []
	private _autoReplyToHeartBeat = true
	public mockConnectCount = 0

	constructor() {
		super()

		// @ts-expect-error this is comparable with ISocketMock
		instances.push(this)

		// Wrap member functions in mocks:
		this.connect = vitest.fn(this.connect)
		this.setEncoding = vitest.fn(this.setEncoding)
		this.destroy = vitest.fn(this.destroy)
		this.pause = vitest.fn(this.pause)
		this.resume = vitest.fn(this.resume)
		this.setTimeout = vitest.fn(this.setTimeout)
		this.setNoDelay = vitest.fn(this.setNoDelay)
		this.setKeepAlive = vitest.fn(this.setKeepAlive)
		this.address = vitest.fn(this.address)
		this.unref = vitest.fn(this.unref)
		this.ref = vitest.fn(this.ref)
		this.end = vitest.fn(this.end)
		this._write = vitest.fn(this._write)
		this.setDefaultEncoding = vitest.fn(this.setDefaultEncoding)
		this._read = vitest.fn(this._read)
		this.read = vitest.fn(this.read)
		this.isPaused = vitest.fn(this.isPaused)
		this.pipe = vitest.fn(this.pipe) as any
		this.unpipe = vitest.fn(this.unpipe)
		this.unshift = vitest.fn(this.unshift)
		this.wrap = vitest.fn(this.wrap)
		this.push = vitest.fn(this.push)
		this._destroy = vitest.fn(this._destroy)
		this._final = vitest.fn(this._final)
		this.cork = vitest.fn(this.cork)
		this.uncork = vitest.fn(this.uncork)

		this.mockSentMessage0 = vitest.fn(this.mockSentMessage0)
		this.mockSentMessage = vitest.fn(this.mockSentMessage)
	}
	static mockClear(): void {
		instances.splice(0, 9999)
	}
	static get instances(): ISocketMock[] {
		return instances
	}

	write(data: unknown, encoding: unknown): boolean {
		this.mockSentMessage0(data, encoding as any)
		return true
	}
	// @ts-expect-error mock
	connect(port: number, host: string): this {
		this.mockConnectCount++
		this.connectedPort = port
		this.connectedHost = host

		return this
	}
	setEncoding(): this {
		return this
	}
	destroy(_error?: Error): this {
		this.destroyed = true
		this.emit('close')
		/* nothing */
		return this
	}
	pause(): this {
		return this
	}
	resume(): this {
		return this
	}
	setTimeout(timeout: number, callback?: (...args: any[]) => void): this {
		if (callback) setTimeout(callback, timeout)
		return this
	}
	setNoDelay(_noDelay?: boolean): this {
		return this
	}
	setKeepAlive(_enable?: boolean, _initialDelay?: number): this {
		return this
	}
	address(): { port: number; family: string; address: string } {
		return { port: 100, family: 'localhost', address: '127.0.0.1' }
	}
	unref(): this {
		return this
	}
	ref(): this {
		return this
	}
	end(): this {
		return this
	}
	_write(_chunk: unknown, _encoding: string, callback: () => void): void {
		callback() /* nothing */
	}
	setDefaultEncoding(_encoding: string): this {
		return this
	}
	_read(_size: number): void {
		// nothing
	}
	read(_size?: number): void {
		// nothing
	}
	isPaused(): boolean {
		return false
	}
	pipe<T extends NodeJS.WritableStream>(destination: T, _options?: { end?: boolean }): T {
		return destination
	}
	unpipe(_destination: NodeJS.WritableStream): this {
		return this
	}
	unshift(_chunk: unknown): void {
		// nothing
	}
	wrap(_oldStream: NodeJS.ReadableStream): this {
		return this
	}
	push(_chunk: unknown, _encoding?: string): boolean {
		return true
	}
	_destroy(): void {
		/* nothing */
	}
	_final(): void {
		/* nothing */
	}
	cork(): void {
		/* nothing */
	}
	uncork(): void {
		/* nothing */
	}

	// ------------------------------------------------------------------------
	// Mock methods:
	mockEmitConnected(): void {
		if (this.connectedPort === 10542) {
			// don't reply on heartbeats on query port
			this._autoReplyToHeartBeat = false
		}

		this.emit('connect')
	}
	mockEmitClose(): void {
		this.emit('close')
	}

	mockSentMessage0(data: unknown, encoding: string): void {
		if (this._autoReplyToHeartBeat) {
			const str: string = typeof data === 'string' ? data : this.decode(data as any)

			if (/<heartbeat>/.exec(str)) {
				try {
					const mosID = (str.match(/<mosID>([^<]+)<\/mosID>/) || [])[1]
					const ncsID = (str.match(/<ncsID>([^<]+)<\/ncsID>/) || [])[1]
					const messageId = getMessageId(str)

					const repl = `<mos>
<mosID>${mosID}</mosID>
<ncsID>${ncsID}</ncsID>
<messageID>${messageId}</messageID>\
  <heartbeat>
    <time>${mosTypes.mosTime.stringify(mosTypes.mosTime.create(Date.now()))}</time>
  </heartbeat>
</mos>\r\n`
					this.mockReceiveMessage(this.encode(repl))
				} catch (e) {
					// eslint-disable-next-line no-console
					console.error('mockReply', str)
					throw e
				}
				return
			}
		}
		this.mockSentMessage(data, encoding)
	}
	mockSentMessage(data: unknown, _encoding: string): void {
		const cb = this._responses.shift()
		if (cb) {
			// send reply:
			setTimeoutOrg(() => {
				Promise.resolve(typeof cb === 'function' ? cb(data) : cb)
					.then((msg) => {
						if (msg) {
							if (Array.isArray(msg)) {
								for (const m of msg) {
									this.mockReceiveMessage(m)
								}
							} else {
								this.mockReceiveMessage(msg)
							}
						}
					})
					// eslint-disable-next-line no-console
					.catch(console.error)
			}, 1)
		}
	}
	mockReceiveMessage(msg: string | Buffer): void {
		this.emit('data', msg)
	}
	mockAddReply(cb: ReplyTypes): void {
		this._responses.push(cb)
	}
	mockClear(): void {
		void this._responses.splice(0, 9999)
		// @ts-expect-error mock hack
		this.mockSentMessage0['mockClear']()
		// @ts-expect-error mock hack
		this.mockSentMessage['mockClear']()
	}
	async mockWaitForSentMessages(): Promise<void> {
		return new Promise<void>((resolve) => {
			const check = () => {
				if (this._responses.length === 0) {
					resolve()
				} else {
					setTimeoutOrg(() => {
						check()
					}, 1)
				}
			}
			check()
		})
	}

	decode(data: Buffer): string {
		return iconv.decode(data, 'utf16-be')
	}
	encode(str: string): Buffer {
		return iconv.encode(str, 'utf16-be')
	}
	setAutoReplyToHeartBeat(autoReplyToHeartBeat: boolean): void {
		this._autoReplyToHeartBeat = autoReplyToHeartBeat
	}
	get autoReplyToHeartBeat(): boolean {
		return this._autoReplyToHeartBeat
	}
}

type SimpleTypes = string | string[] | false | Buffer | Buffer[]

type ReplyTypes =
	| SimpleTypes
	| ((data: any) => SimpleTypes)
	| Promise<SimpleTypes>
	| ((data: any) => Promise<SimpleTypes>)

export type ISocketMock = Mocked<SocketMock>
