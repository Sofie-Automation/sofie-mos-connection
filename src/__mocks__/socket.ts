/* eslint-disable @typescript-eslint/unbound-method */
import { EventEmitter } from 'events'
import { Socket } from 'net'
import { MosTime } from '../dataTypes/mosTime'

import * as iconv from 'iconv-lite'
import { getMessageId } from '../__tests__/lib'
iconv.encodingExists('utf16-be')

// Mock the Socket class in 'net':
const setTimeoutOrg = setTimeout

const instances: Array<ISocketMock> = []

export class SocketMock extends EventEmitter implements Socket {
	// Mock implementation:
	bufferSize = 0
	remoteAddress = ''
	remoteFamily = ''
	remotePort = 0
	localAddress = ''
	localPort = 0
	bytesRead = 0
	bytesWritten = 0
	connecting = false
	destroyed = false
	writable = false
	readable = false

	readonly readableEncoding: BufferEncoding | null = null
	readonly readableEnded: boolean = false
	readonly readableFlowing: boolean | null = null

	readonly writableEnded: boolean = false
	readonly writableFinished: boolean = false
	readonly writableHighWaterMark: number = 0
	readonly writableLength: number = 0
	readonly writableObjectMode: boolean = false

	readonly readableHighWaterMark: number = 0
	readonly readableLength: number = 0
	readonly readableObjectMode: boolean = false

	public name = ''
	public connectedPort = 0
	public connectedHost = ''

	// [Symbol.asyncIterator](): AsyncIterableIterator<any>

	private _responses: Array<ReplyTypes> = []
	private _autoReplyToHeartBeat = true

	constructor() {
		super()

		// @ts-expect-error this is comparable with ISocketMock
		instances.push(this)

		// Wrap member functions in mocks:
		this.connect = jest.fn(this.connect)
		this.setEncoding = jest.fn(this.setEncoding)
		this.destroy = jest.fn(this.destroy)
		this.pause = jest.fn(this.pause)
		this.resume = jest.fn(this.resume)
		this.setTimeout = jest.fn(this.setTimeout)
		this.setNoDelay = jest.fn(this.setNoDelay)
		this.setKeepAlive = jest.fn(this.setKeepAlive)
		this.address = jest.fn(this.address)
		this.unref = jest.fn(this.unref)
		this.ref = jest.fn(this.ref)
		this.end = jest.fn(this.end)
		this._write = jest.fn(this._write)
		this.setDefaultEncoding = jest.fn(this.setDefaultEncoding)
		this._read = jest.fn(this._read)
		this.read = jest.fn(this.read)
		this.isPaused = jest.fn(this.isPaused)
		this.pipe = jest.fn(this.pipe)
		this.unpipe = jest.fn(this.unpipe)
		this.unshift = jest.fn(this.unshift)
		this.wrap = jest.fn(this.wrap)
		this.push = jest.fn(this.push)
		this._destroy = jest.fn(this._destroy)
		this._final = jest.fn(this._final)
		this.cork = jest.fn(this.cork)
		this.uncork = jest.fn(this.uncork)

		this.mockSentMessage0 = jest.fn(this.mockSentMessage0)
		this.mockSentMessage = jest.fn(this.mockSentMessage)
	}
	static mockClear(): void {
		instances.splice(0, 9999)
	}
	static get instances(): ISocketMock[] {
		return instances
	}

	write(data: unknown, encoding: unknown): boolean {
		this.mockSentMessage0(data, encoding as any)
		return true
	}
	// @ts-expect-error mock
	connect(port: number, host: string): SocketMock {
		this.connectedPort = port
		this.connectedHost = host

		if (this.connectedPort === 10542) {
			// don't reply on heartbeats on query port
			this._autoReplyToHeartBeat = false
		}

		this.emit('connect')
		return this
	}
	setEncoding(): this {
		return this
	}
	destroy(_error?: Error): this {
		this.destroyed = true
		this.emit('close')
		/* nothing */
		return this
	}
	pause(): this {
		return this
	}
	resume(): this {
		return this
	}
	setTimeout(timeout: number, callback?: (...args: any[]) => void): this {
		if (callback) setTimeout(callback, timeout)
		return this
	}
	setNoDelay(_noDelay?: boolean): this {
		return this
	}
	setKeepAlive(_enable?: boolean, _initialDelay?: number): this {
		return this
	}
	address(): { port: number; family: string; address: string } {
		return { port: 100, family: 'localhost', address: '127.0.0.1' }
	}
	unref(): this {
		return this
	}
	ref(): this {
		return this
	}
	end(): this {
		return this
	}
	_write(_chunk: unknown, _encoding: string, callback: () => void): void {
		callback() /* nothing */
	}
	setDefaultEncoding(_encoding: string): this {
		return this
	}
	_read(_size: number): void {
		// nothing
	}
	read(_size?: number): void {
		// nothing
	}
	isPaused(): boolean {
		return false
	}
	pipe<T extends NodeJS.WritableStream>(destination: T, _options?: { end?: boolean }): T {
		return destination
	}
	unpipe(_destination: NodeJS.WritableStream): this {
		return this
	}
	unshift(_chunk: unknown): void {
		// nothing
	}
	wrap(_oldStream: NodeJS.ReadableStream): this {
		return this
	}
	push(_chunk: unknown, _encoding?: string): boolean {
		return true
	}
	_destroy(): void {
		/* nothing */
	}
	_final(): void {
		/* nothing */
	}
	cork(): void {
		/* nothing */
	}
	uncork(): void {
		/* nothing */
	}

	// ------------------------------------------------------------------------
	// Mock methods:
	mockSentMessage0(data: unknown, encoding: string): void {
		if (this._autoReplyToHeartBeat) {
			const str: string = typeof data === 'string' ? data : this.decode(data as any)

			if (str.match(/<heartbeat>/)) {
				try {
					const mosID = (str.match(/<mosID>([^<]+)<\/mosID>/) || [])[1]
					const ncsID = (str.match(/<ncsID>([^<]+)<\/ncsID>/) || [])[1]
					const messageId = getMessageId(str)

					const repl = `<mos>
<mosID>${mosID}</mosID>
<ncsID>${ncsID}</ncsID>
<messageID>${messageId}</messageID>\
  <heartbeat>
    <time>${new MosTime().toString()}</time>
  </heartbeat>
</mos>\r\n`
					this.mockReceiveMessage(this.encode(repl))
				} catch (e) {
					console.error('mockReply', str)
					throw e
				}
				return
			}
		}
		this.mockSentMessage(data, encoding)
	}
	mockSentMessage(data: unknown, _encoding: string): void {
		const cb = this._responses.shift()
		if (cb) {
			// send reply:
			setTimeoutOrg(() => {
				Promise.resolve(typeof cb === 'function' ? cb(data) : cb)
					.then((msg) => {
						if (msg) {
							if (Array.isArray(msg)) {
								for (const m of msg) {
									this.mockReceiveMessage(m)
								}
							} else {
								this.mockReceiveMessage(msg)
							}
						}
					})
					.catch(console.error)
			}, 1)
		}
	}
	mockReceiveMessage(msg: string | Buffer): void {
		this.emit('data', msg)
	}
	mockAddReply(cb: ReplyTypes): void {
		this._responses.push(cb)
	}
	mockClear(): void {
		this._responses.splice(0, 9999)
		// @ts-expect-error mock hack
		this.mockSentMessage0['mockClear']()
		// @ts-expect-error mock hack
		this.mockSentMessage['mockClear']()
	}
	async mockWaitForSentMessages(): Promise<void> {
		return new Promise<void>((resolve) => {
			const check = () => {
				if (this._responses.length === 0) {
					resolve()
				} else {
					setTimeoutOrg(() => {
						check()
					}, 1)
				}
			}
			check()
		})
	}

	decode(data: Buffer): string {
		return iconv.decode(data, 'utf16-be')
	}
	encode(str: string): Buffer {
		return iconv.encode(str, 'utf16-be')
	}
	setAutoReplyToHeartBeat(autoReplyToHeartBeat: boolean): void {
		this._autoReplyToHeartBeat = autoReplyToHeartBeat
	}
}

type SimpleTypes = string | string[] | false | Buffer | Buffer[]

type ReplyTypes =
	| SimpleTypes
	| ((data: any) => SimpleTypes)
	| Promise<SimpleTypes>
	| ((data: any) => Promise<SimpleTypes>)

export type ISocketMock = jest.Mocked<SocketMock>
