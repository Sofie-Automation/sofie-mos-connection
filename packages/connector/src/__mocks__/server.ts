/* eslint-disable @typescript-eslint/unbound-method */
import { EventEmitter } from 'events'
import { Server } from 'net'
import { SocketMock } from './socket'

import * as iconv from 'iconv-lite'
iconv.encodingExists('utf16-be')

// Mock the Server class in 'net':
const setTimeoutOrg = setTimeout

const instances: Array<IServerMock> = []

export class ServerMock extends EventEmitter implements Server {
	// Mock implementation:

	maxConnections = 0
	connections = 0
	listening = false
	listenToPort = 0

	private _responses: Array<(data: any) => string | Buffer> = []

	constructor() {
		super()

		// @ts-expect-error this is comparable with ISocketMock
		instances.push(this)

		this.listen = jest.fn(this.listen)
		this.close = jest.fn(this.close)
		this.address = jest.fn(this.address)
		this.getConnections = jest.fn(this.getConnections)
		this.ref = jest.fn(this.ref)
		this.unref = jest.fn(this.unref)
	}
	static mockClear(): void {
		instances.splice(0, 9999)
	}
	static get instances(): IServerMock[] {
		return instances
	}

	// @ts-expect-error mock hack
	listen(port: number): this {
		this.listenToPort = port
		setTimeoutOrg(() => {
			this.listening = true
			this.emit('listening', {})
		}, 1)
		return this
	}
	close(callback?: () => void): this {
		if (callback) callback()
		setTimeoutOrg(() => {
			this.emit('close')
		}, 1)
		return this
	}
	address(): { port: number; family: string; address: string } {
		return { port: 0, family: 'string', address: 'string' }
	}
	getConnections(cb: (error: Error | null, count: number) => void): this {
		cb(null, 0)
		return this
	}
	ref(): this {
		return this
	}
	unref(): this {
		return this
	}

	// Mock methods:
	mockNewConnection(): SocketMock {
		// "Someone has connected"
		const socket = new SocketMock()
		socket.on('data', (d) => {
			this.emit('data', d)
		})

		this.emit('connection', socket)

		return socket
	}

	mockSentMessage(data: string, _encoding: unknown): void {
		const cb = this._responses.shift()
		if (cb) {
			// send reply:
			let msg

			setTimeoutOrg(() => {
				if (typeof cb === 'string') {
					msg = cb
				} else {
					msg = cb(data)
				}

				this.mockReceiveMessage(msg)
			}, 1)
		}
	}
	mockReceiveMessage(msg: string | Buffer): void {
		this.emit('data', msg)
	}
	mockAddReply(cb: (data: any) => string | Buffer): void {
		this._responses.push(cb)
	}
	mockClear(): void {
		this._responses.splice(0, 9999)
	}
	decode(data: Buffer): string {
		return iconv.decode(data, 'utf16-be')
	}
	encode(str: string): Buffer {
		return iconv.encode(str, 'utf16-be')
	}
}

export type IServerMock = jest.Mocked<ServerMock>
