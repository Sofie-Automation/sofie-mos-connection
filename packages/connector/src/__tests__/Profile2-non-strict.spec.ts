import {
	checkReplyToServer,
	clearMocks,
	decode,
	doBeforeAll,
	encode,
	fakeIncomingMessage,
	fixSnapshot,
	getMessageId,
	getMosConnection,
	getMosDevice,
	getXMLReply,
	mosTypes,
	setupMocks,
} from './lib.js'
import {
	MosConnection,
	MosDevice,
	IMOSObject,
	IMOSItem,
	IMOSItemAction,
	IMOSItemStatus,
	IMOSROAck,
	IMOSROAction,
	IMOSROReadyToAir,
	IMOSROStory,
	IMOSRunningOrder,
	IMOSRunningOrderBase,
	IMOSRunningOrderStatus,
	IMOSStoryAction,
	IMOSStoryStatus,
	IMOSListMachInfo,
	IMOSString128,
} from '../index.js'
import { SocketMock } from '../__mocks__/socket.js'
import { ServerMock } from '../__mocks__/server.js'
import { xmlData, xmlApiData } from '../__mocks__/testData.js'
import { describe, test, expect, beforeAll, beforeEach, afterAll, vitest, MockedFunction } from 'vitest'

/* eslint-disable @typescript-eslint/no-unused-vars */
// @ts-ignore imports are unused
import { Socket } from 'net'
/* eslint-enable @typescript-eslint/no-unused-vars */

beforeAll(() => {
	setupMocks()
})
beforeEach(() => {
	clearMocks()
})
describe('Profile 2 - non strict', () => {
	let mosDevice: MosDevice
	let mosConnection: MosConnection
	let socketMockLower: SocketMock
	let socketMockUpper: SocketMock

	let serverSocketMockLower: SocketMock
	let serverSocketMockUpper: SocketMock
	let serverSocketMockQuery: SocketMock

	let onRequestMachineInfo: MockedFunction<any>
	let onRequestMOSObject: MockedFunction<any>
	let onRequestAllMOSObjects: MockedFunction<any>
	let onCreateRunningOrder: MockedFunction<any>
	let onReplaceRunningOrder: MockedFunction<any>
	let onDeleteRunningOrder: MockedFunction<any>
	let onRequestRunningOrder: MockedFunction<any>
	let onMetadataReplace: MockedFunction<any>
	let onRunningOrderStatus: MockedFunction<any>
	let onStoryStatus: MockedFunction<any>
	let onItemStatus: MockedFunction<any>
	let onReadyToAir: MockedFunction<any>
	let onROInsertStories: MockedFunction<any>
	let onROInsertItems: MockedFunction<any>
	let onROReplaceStories: MockedFunction<any>
	let onROReplaceItems: MockedFunction<any>
	let onROMoveStories: MockedFunction<any>
	let onROMoveItems: MockedFunction<any>
	let onRODeleteStories: MockedFunction<any>
	let onRODeleteItems: MockedFunction<any>
	let onROSwapStories: MockedFunction<any>
	let onROSwapItems: MockedFunction<any>

	const mockReplyRoAck = vitest.fn((data) => {
		const str = decode(data)
		const messageID = getMessageId(str)
		return encode(getXMLReply(messageID, xmlData.roAck))
	})

	beforeAll(async () => {
		SocketMock.mockClear()
		ServerMock.mockClear()

		mosConnection = await getMosConnection(
			{
				'0': true,
				'1': true,
				'2': true,
			},
			false
		)
		mosDevice = await getMosDevice(mosConnection)

		// Profile 0:
		onRequestMachineInfo = vitest.fn(async () => {
			return xmlApiData.machineInfo
		})
		mosDevice.onRequestMachineInfo(async (): Promise<IMOSListMachInfo> => {
			return onRequestMachineInfo()
		})
		// Profile 1:
		onRequestMOSObject = vitest.fn()
		onRequestAllMOSObjects = vitest.fn()
		mosDevice.onRequestMOSObject(async (objId: string): Promise<IMOSObject | null> => {
			return onRequestMOSObject(objId)
		})
		mosDevice.onRequestAllMOSObjects(async (): Promise<Array<IMOSObject>> => {
			return onRequestAllMOSObjects()
		})

		// Profile 2:
		const roAckReply = async () => {
			const ack: IMOSROAck = {
				ID: mosTypes.mosString128.create('runningOrderId'),
				Status: mosTypes.mosString128.create('OK'),
				Stories: [],
			}
			return ack
		}
		onCreateRunningOrder = vitest.fn(roAckReply)
		onReplaceRunningOrder = vitest.fn(roAckReply)
		onDeleteRunningOrder = vitest.fn(roAckReply)
		onRequestRunningOrder = vitest.fn(async () => {
			return xmlApiData.roCreate
		})
		onMetadataReplace = vitest.fn(roAckReply)
		onRunningOrderStatus = vitest.fn(roAckReply)
		onStoryStatus = vitest.fn(roAckReply)
		onItemStatus = vitest.fn(roAckReply)
		onReadyToAir = vitest.fn(roAckReply)
		onROInsertStories = vitest.fn(roAckReply)
		onROInsertItems = vitest.fn(roAckReply)
		onROReplaceStories = vitest.fn(roAckReply)
		onROReplaceItems = vitest.fn(roAckReply)
		onROMoveStories = vitest.fn(roAckReply)
		onROMoveItems = vitest.fn(roAckReply)
		onRODeleteStories = vitest.fn(roAckReply)
		onRODeleteItems = vitest.fn(roAckReply)
		onROSwapStories = vitest.fn(roAckReply)
		onROSwapItems = vitest.fn(roAckReply)

		mosDevice.onCreateRunningOrder(async (ro: IMOSRunningOrder): Promise<IMOSROAck> => {
			return onCreateRunningOrder(ro)
		})
		mosDevice.onReplaceRunningOrder(async (ro: IMOSRunningOrder): Promise<IMOSROAck> => {
			return onReplaceRunningOrder(ro)
		})
		mosDevice.onDeleteRunningOrder(async (runningOrderId: IMOSString128): Promise<IMOSROAck> => {
			return onDeleteRunningOrder(runningOrderId)
		})
		mosDevice.onRequestRunningOrder(async (runningOrderId: IMOSString128): Promise<IMOSRunningOrder | null> => {
			return onRequestRunningOrder(runningOrderId)
		})
		mosDevice.onMetadataReplace(async (metadata: IMOSRunningOrderBase): Promise<IMOSROAck> => {
			return onMetadataReplace(metadata)
		})
		mosDevice.onRunningOrderStatus(async (status: IMOSRunningOrderStatus): Promise<IMOSROAck> => {
			return onRunningOrderStatus(status)
		})
		mosDevice.onStoryStatus(async (status: IMOSStoryStatus): Promise<IMOSROAck> => {
			return onStoryStatus(status)
		})
		mosDevice.onItemStatus(async (status: IMOSItemStatus): Promise<IMOSROAck> => {
			return onItemStatus(status)
		})
		mosDevice.onReadyToAir(async (Action: IMOSROReadyToAir): Promise<IMOSROAck> => {
			return onReadyToAir(Action)
		})
		mosDevice.onROInsertStories(
			async (Action: IMOSStoryAction, Stories: Array<IMOSROStory>): Promise<IMOSROAck> => {
				return onROInsertStories(Action, Stories)
			}
		)
		mosDevice.onROInsertItems(async (Action: IMOSItemAction, Items: Array<IMOSItem>): Promise<IMOSROAck> => {
			return onROInsertItems(Action, Items)
		})
		mosDevice.onROReplaceStories(
			async (Action: IMOSStoryAction, Stories: Array<IMOSROStory>): Promise<IMOSROAck> => {
				return onROReplaceStories(Action, Stories)
			}
		)
		mosDevice.onROReplaceItems(async (Action: IMOSItemAction, Items: Array<IMOSItem>): Promise<IMOSROAck> => {
			return onROReplaceItems(Action, Items)
		})
		mosDevice.onROMoveStories(
			async (Action: IMOSStoryAction, Stories: Array<IMOSString128>): Promise<IMOSROAck> => {
				return onROMoveStories(Action, Stories)
			}
		)
		mosDevice.onROMoveItems(async (Action: IMOSItemAction, Items: Array<IMOSString128>): Promise<IMOSROAck> => {
			return onROMoveItems(Action, Items)
		})
		mosDevice.onRODeleteStories(async (Action: IMOSROAction, Stories: Array<IMOSString128>): Promise<IMOSROAck> => {
			return onRODeleteStories(Action, Stories)
		})
		mosDevice.onRODeleteItems(async (Action: IMOSStoryAction, Items: Array<IMOSString128>): Promise<IMOSROAck> => {
			return onRODeleteItems(Action, Items)
		})
		mosDevice.onROSwapStories(
			async (Action: IMOSROAction, StoryID0: IMOSString128, StoryID1: IMOSString128): Promise<IMOSROAck> => {
				return onROSwapStories(Action, StoryID0, StoryID1)
			}
		)
		mosDevice.onROSwapItems(
			async (Action: IMOSStoryAction, ItemID0: IMOSString128, ItemID1: IMOSString128): Promise<IMOSROAck> => {
				return onROSwapItems(Action, ItemID0, ItemID1)
			}
		)
		const b = doBeforeAll()
		socketMockLower = b.socketMockLower
		socketMockUpper = b.socketMockUpper
		serverSocketMockLower = b.serverSocketMockLower
		serverSocketMockUpper = b.serverSocketMockUpper
		serverSocketMockQuery = b.serverSocketMockQuery

		mosConnection.checkProfileValidness()
		mosDevice.checkProfileValidness()
	})
	beforeEach(() => {
		onRequestMOSObject.mockClear()
		onRequestAllMOSObjects.mockClear()

		onCreateRunningOrder.mockClear()
		onReplaceRunningOrder.mockClear()
		onDeleteRunningOrder.mockClear()
		onRequestRunningOrder.mockClear()
		onMetadataReplace.mockClear()
		onRunningOrderStatus.mockClear()
		onStoryStatus.mockClear()
		onItemStatus.mockClear()
		onReadyToAir.mockClear()
		onROInsertStories.mockClear()
		onROInsertItems.mockClear()
		onROReplaceStories.mockClear()
		onROReplaceItems.mockClear()
		onROMoveStories.mockClear()
		onROMoveItems.mockClear()
		onRODeleteStories.mockClear()
		onRODeleteItems.mockClear()
		onROSwapStories.mockClear()
		onROSwapItems.mockClear()

		socketMockLower.mockClear()
		socketMockUpper.mockClear()
		serverSocketMockLower.mockClear()
		serverSocketMockUpper.mockClear()
		serverSocketMockQuery.mockClear()

		mockReplyRoAck.mockClear()
	})
	afterAll(async () => {
		await mosDevice.dispose()
		await mosConnection.dispose()
	})
	describe('deprecated messages', () => {
		// These methods are still supported, but will be removed in future versions of the mos protocol
		test('roStoryMove - missing second <storyID>', async () => {
			// Note: from documentation:
			// https://mosprotocol.com/wp-content/MOS-Protocol-Documents/MOS_Protocol_Version_2.8.5_Final.htm#roStoryMove
			// **Note**: If the second <storyID> tag is blank move the story to the bottom of the Running Order.

			// Fake incoming message on socket:
			const messageId = await fakeIncomingMessage(serverSocketMockLower, xmlData.roStoryMove_offspec_missing)
			expect(onROMoveStories).toHaveBeenCalledTimes(1)
			expect(onROMoveStories.mock.calls[0][0]).toEqual(xmlApiData.roElementAction_roStoryMove_blank_action)
			expect(onROMoveStories.mock.calls[0][1]).toEqual(xmlApiData.roElementAction_roStoryMove_stories)
			expect(fixSnapshot(onROMoveStories.mock.calls)).toMatchSnapshot()
			await checkReplyToServer(serverSocketMockLower, messageId, '<roAck>')
		})
		test('roStoryMoveMultiple with single storyId', async () => {
			// Fake incoming message on socket:
			const messageId = await fakeIncomingMessage(
				serverSocketMockLower,
				xmlData.roStoryMoveMultiple_single_storyId
			)
			expect(onROMoveStories).toHaveBeenCalledTimes(1)
			expect(onROMoveStories.mock.calls[0][0]).toEqual(
				xmlApiData.roElementAction_roStoryMoveMultiple_single_storyId_offspec_action
			)
			expect(onROMoveStories.mock.calls[0][1]).toEqual(
				xmlApiData.roElementAction_roStoryMoveMultiple_single_storyId_offspec_stories
			)
			expect(fixSnapshot(onROMoveStories.mock.calls)).toMatchSnapshot()
			await checkReplyToServer(serverSocketMockLower, messageId, '<roAck>')
		})
	})
})
