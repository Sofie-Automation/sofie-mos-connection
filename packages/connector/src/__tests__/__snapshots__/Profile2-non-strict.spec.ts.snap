// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Profile 2 - non strict > deprecated messages > roStoryMove - missing second <storyID> > <mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>1633</messageID>
  <roAck>
    <roID>runningOrderId</roID>
    <roStatus>OK</roStatus>
  </roAck>
</mos> 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>1633</messageID>
  <roAck>
    <roID>runningOrderId</roID>
    <roStatus>OK</roStatus>
  </roAck>
</mos>"
`;

exports[`Profile 2 - non strict > deprecated messages > roStoryMove - missing second <storyID> 1`] = `
[
  [
    {
      "RunningOrderID": {
        "_mosString128": "5PM",
      },
      "StoryID": {
        "_mosString128": "",
      },
    },
    [
      {
        "_mosString128": "V: BRIDGE COLLAPSE",
      },
    ],
  ],
]
`;

exports[`Profile 2 - non strict > deprecated messages > roStoryMoveMultiple with single storyId > <mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>1634</messageID>
  <roAck>
    <roID>runningOrderId</roID>
    <roStatus>OK</roStatus>
  </roAck>
</mos> 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>1634</messageID>
  <roAck>
    <roID>runningOrderId</roID>
    <roStatus>OK</roStatus>
  </roAck>
</mos>"
`;

exports[`Profile 2 - non strict > deprecated messages > roStoryMoveMultiple with single storyId 1`] = `
[
  [
    {
      "RunningOrderID": {
        "_mosString128": "5PM",
      },
      "StoryID": {
        "_mosString128": "",
      },
    },
    [
      {
        "_mosString128": "2",
      },
    ],
  ],
]
`;
